maildir/maildir.c

```diff
@@ -1610,6 +1612,57 @@ static enum MailboxType maildir_path_probe(const char *path, const struct stat *
   return MUTT_UNKNOWN;
 }
 
+/**
+ * maildir_check_dir - create a new maildir directory tree
+ * @param path Path to create the maildir under
+ * @retval -1 Creation failed, we attempt to cleanup anything we already created
+ * @retval  0 Creation success
+ *
+ */
+static int maildir_create_maildir(const char *path)
+{
+  char tmp[PATH_MAX];
+
+  for (int i = 0; i < mutt_array_size(maildir_subdirs); i++)
+  {
+    snprintf(tmp, sizeof(tmp), "%s/%s", path, maildir_subdirs[i]);
+
+    if (mutt_file_mkdir(tmp, S_IRWXU) != 0)
+    {
+      mutt_perror(tmp);
+
+      /* Cleanup what we've already made */
+      for (int j = i - 1; j >= 0; j--)
+      {
+        snprintf(tmp, sizeof(tmp), "%s/%s", path, maildir_subdirs[j]);
+        /* Should probably check errors again here... */
+        /* alert user if something went wrong */
+        rmdir(tmp);
+      }
+      rmdir(path);
+      return -1;
+    }
+  }
+  return 0;
+}
+
+/**
+ *  maildir_create_mailbox - Create a maildir mailbox - Implements - MxOps::mbox_create()
+ */
+struct Mailbox *maildir_create_mailbox(struct Account *a, const char *path, const char *name)
+{
+  if (maildir_create_maildir(path) != 0)
+    return NULL;
+  struct Mailbox *m = mailbox_new();
+
+  mutt_buffer_strcpy(&m->pathbuf, path);
+  m->realpath = mutt_str_dup(path);
```

Comment:
I'm not sure I understand the difference in `pathbuf` vs `realpath`?

Reply:
> I'm not sure I understand the difference in pathbuf vs realpath?

I'm also confused about the difference. In `mailbox.h`, `pathbuf` is the only undocumented field and `realpath`'s documentation is its usage, not what it is.

Is the difference that one is fully-qualified and one isn't?

Reply:
Originally it was just `path`, a big char array.

Then `realpath` came along to store the 'canonical' path.
In local terms that means following symlinks, for network paths the full `proto://user@host.com/path`.

To avoid lots of large char arrays, upstream replaced `path` with `pathbuf` (`struct Buffer`).

The other thing in the mix, is the descriptive name, for `named-mailboxes`.

---

I spent several months working out a better scheme for the paths, see #2104
It documents how paths should be handled (it creates a Path API).
Unfortunately, I couldn't figure out how to fit it into the existing mailboxes without breaking things.
Now that we're busy breaking things, it might be time to revisit paths.